### PR TITLE
Feature/coopsigsave

### PR DESCRIPTION
**Pull request by mezdahun**
In this PR we create the first working prototype of the replay tool with project specific saved files and we define what needs to be saved when the APP_VERSION is CooperativeSignalling.

1. Added meter value and signaling for saving in monitoring/ifdb as new cs specific methods
2. These are called in cs_sims so cs specific attributes are saved
3. These are zarr files and other formats are not implemented
4. These are then loaded back with the DataLoader class that can now load back cs specific zarr files
5. The loaded data is used to instanciate an ExperimentReplay class instance that can visualize what happened in the played experiment (using now cs specific zarr files)
6. The version of the app is controlled with the "APP_VESRION" env flag

To test the method: 
1. create a "test.env" empty file in the root ABM folder (this will be overwritten)
2. run `EXPERIMENT_NAME=test python /path/to/abm/ABM/abm/data/metaprotocol/experiments/archive/CooperativeSignalling/template/ExperimentCoopSigTemplate.py`
3. this will generate experimental data in data/simulation_data/test
4. change the path in ABM/load_data_test.py to the absolute path of the generated test folder
5. the data runs are joined and summarized into a single summary folder under test and a replay window opens up.

Note that most of the functionalities are not yet implemented in the replay tool to handle cs specific stuff yet, only the visualization of the agents.

### CODE REVIEW

**Comment by mezdahun** [comment]
https://user-images.githubusercontent.com/38374698/215784656-8f8deb88-0113-411e-9390-10633ebbed40.mp4

**Comment by mezdahun** [comment]
@vagechirkov This is just an example of the whole cycle how we can include arbitrary parameters to be saved and use our already created infrastructure to replay these. I would be happy if we could continue on this branch instead of you implementing this again from scratch.

**Comment by mezdahun** [comment]
https://github.com/influxdata/telegraf/issues/12563

**Comment by mezdahun** [comment]
Collected resources are now saved under the same name as with the base version, so all functionalities of the replay tool are now usable with data saved from CS version. New features that could be implemented:
e.g: Time spent with signalling (same way as currently the time spent in relocation mode)

I thin this can be left after/during our refactoring sprint. The next step would be to define our experiment files similarly to https://github.com/scioip34/ABM/blob/feature/coopsigsave/abm/data/metaprotocol/experiments/archive/CooperativeSignalling/template/ExperimentCoopSigTemplate.py

@vagechirkov @DominikDeffner I have lost my notes about the planned experiments, can we collect this here and then define the experiment files accordingly?

For example:
1.) Exp1: Showing the effect of signalling in collective efficiency:
  - fix arena size, agent velocity, agent size, taxis behavior
  - change patch velocity gradually
  - agents a.) never signal, b.) always signal

**Comment by mezdahun** [comment]
Added first sharp experiment as in previous comment under this file: https://github.com/scioip34/ABM/blob/feature/coopsigsave/abm/data/metaprotocol/experiments/archive/CooperativeSignalling/CSExp1.py

**Comment by vagechirkov** [comment]
Hey @mezdahun, sorry for the delay, I'll review the PR tomorrow! I think we discussed the following combinations:

1. `Resource speed` vs. `resource size ratio` for `no signaling` and `all signaling` conditions. I think we can do it for the whole range of `signaling propensity` (i.e., `0.0`;`0.1`; `0.2`; `0.3` ... `0.9`;`1.0`).
2. `Resource speed` vs. `memory size` for `all signaling` condition.

**Comment by mezdahun** [comment]
Sounds good and no worries, I just had these changes in mind already and was faster for me to quickly have a look. Can you formulize these in terms of experiment files similarly to https://github.com/scioip34/ABM/blob/feature/coopsigsave/abm/data/metaprotocol/experiments/archive/CooperativeSignalling/CSExp1.py ? Then I will have a look and we can run them.

**Comment by vagechirkov** [comment]
@mezdahun could you please clarify what should be in this line for different experiment files?

```python

EXP_NAME = os.getenv("EXPERIMENT_NAME", "")

```

**Comment by vagechirkov** [comment]
@mezdahun Another quick question. I'm wondering how we convert simulation time into real-time units. I need this to make the replay tool compatible with the experimental data.

**Comment by mezdahun** [comment]
> @mezdahun could you please clarify what should be in this line for different experiment files?
> 
> ```python
> EXP_NAME = os.getenv("EXPERIMENT_NAME", "")
> ```

This should be kept as is. The `EXPERIMENT_NAME` is read from the environment variables that is passed to the singularity container on the cluster. However you name your experiment file (e.g. CSExpN.py) the filename will be this EXPERIMENT_NAME. Also, the filename should not contain any underscore (_) characters as this has a meaning in the sbatch script that handles the `.env` files.

**Comment by mezdahun** [comment]
> @mezdahun Another quick question. I'm wondering how we convert simulation time into real-time units. I need this to make the replay tool compatible with the experimental data.

To match this we will have to match the length scale of the 2 systems first. You can say that e.g. 1 unit in Unity (e.g. meter) is 15 units in this framework (as in 15 px). Then we match time scale. What is the velocity of the agent in Unity and then match that with this. For example if the agent in unity has a speed of 0.3m/s, where the unit timescale of the Unity system is seconds, we say that means 5px/"second" in our system. Our agent has a speed of 1px/timestep, that means 1second in unity corresponds to 5 timesteps in ABM world. Alternatively you could match the timescale first and adjust everything else accordingly it is pretty arbitrary at this point.

**Comment by vagechirkov** [comment]
All right, thanks for the explanation! Let's figure out and define the conversion procedure!

Current environment setting in unity:

- Arena size is ~375 unity meters (don't ask why, some random design choices 😅). We can freely scale the size, but it might change the look of the crater 🌚
- Agent speed is 7 unity meters per second. To make the interaction smoother, when participant presses the movement key, avatar's speed is increasing according to the force of button press (within ~400ms). The same on the button release. But I think this should not change the average speed significantly.
- Agent rotation speed is 150 deg/s.

**Comment by vagechirkov** [comment]
Let's set a unity meter to 10 pixels? Then the arena should be 3750 pixels and the agent speed should be 70 pixels/s, right?

**Comment by vagechirkov** [comment]
Hi, I'm a little unsure how we should proceed with this PR. Should we keep working on the replay tool here? @mezdahun

**Comment by mezdahun** [comment]
In https://github.com/scioip34/ABM/pull/64/commits/05553498840efaebbe5d0b794867c3d7cd59eaef the following changes has been done:
- the FOV has been changed to 0.5 as in real experiments
- I used the following rescaling assumptions: 1 unity m = 3 px, and 1 unity second = 2 simulation ts. This way agent velocity is 10px/ts ~= 7m/s, max turning rate 0.21rad/ts =~150deg/sec (except in taxis, where this number is doubled as our agents can not stop in contrast to real ones), arena size 1200px =~ 400m
- resource speed and size is given w.r.t. agent size
- @vagechirkov @DominikDeffner please check this and give me a green light so I can start running this on the cluster

**Comment by DominikDeffner** [comment]
Sounds all good to me, but better wait for @vagechirkov to confirm!

**Comment by vagechirkov** [comment]
Sorry, I need some time to double check parameters in Unity!

**Comment by vagechirkov** [comment]
Regarding FOV, here is the camera settings in Unity: 


<details>

image
image 

From Unity Manual: https://docs.unity3d.com/Manual/class-Camera.html
image


</details>


So think we should set FOV=0.25

**Comment by vagechirkov** [comment]
> I used the following rescaling assumptions: 1 unity m = 3 px, and 1 unity second = 2 simulation ts.

I think this should be fine. In the experiment results we will have 10 samples per second.

**Comment by vagechirkov** [comment]
Other settings look good! Thank you @mezdahun!

**Comment by mezdahun** [comment]
Awesome! I changed the FOV, and will now start the experiment!

File: abm/data/metaprotocol/experiments/archive/CooperativeSignalling/CSExp4.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+    # Memory
+    Constant("MEMORY_DEPTH", 1),
+    # Vision
+    Constant("AGENT_FOV", 0.25),  # unlimited FOV
+    # Resource
+    Constant("RES_THETA", 0.2),
+    Tunable("RES_VEL", values_override=resource_vels),
+    Constant("METER_TO_RES_MULTI", 1),
+    Constant("DETECTION_RANGE", detection_range),
+    Constant("T", 250),
```
Comment (mezdahun) [inline abm/data/metaprotocol/experiments/archive/CooperativeSignalling/CSExp4.py]: this is equivalent to 125 real seconds ~ 2 minutes

Comment (vagechirkov) [reply abm/data/metaprotocol/experiments/archive/CooperativeSignalling/CSExp4.py]: Oops, I missed that! Thank you!

**Comment by vagechirkov** [comment]
I have this error when I try to replay the last simulation: 

image

For some reason it doesn't seem to like when there is only one agent 🤷‍♂️

@mezdahun Any tips for a quick fix? 🙂 I can also skip the batches with only one agent for now

**Comment by vagechirkov** [comment]
> I have this error when I try to replay the last simulation:
> 
> image
> 
> For some reason it doesn't seem to like when there is only one agent 🤷‍♂️
> 
> @mezdahun Any tips for a quick fix? 🙂 I can also skip the batches with only one agent for now

It works after I commented this part of the code 😅

**Pull request merged**